src/mlpack/methods/rbm/spike_slab_layer_impl.hpp

```diff
@@ -0,0 +1,145 @@
+using namespace mlpack::distribution;
+
+namespace mlpack {
+namespace rbm { /** Artificial Neural Network. */
+
+template<typename InputDataType, typename OutputDataType>
+SpikeSlabLayer<InputDataType, OutputDataType>::SpikeSlabLayer(
+    const size_t inSize,
+    const size_t outSize,
+    const size_t poolSize,
+    const double radius,
+    const bool typeVisible):
+    inSize(inSize),
+    outSize(outSize),
+    poolSize(poolSize),
+    radius(radius),
+    typeVisible(typeVisible)
+{
+  weights.set_size(
+      inSize * ((outSize * inSize) + (poolSize * poolSize) + 1) + outSize,
+      outSize * outSize);
+}
+
+template<typename InputDataType, typename OutputDataType>
+void SpikeSlabLayer<InputDataType, OutputDataType>::Reset()
+{
+  if (typeVisible)
+  {
+    visibleSize = inSize;
+    hiddenSize = outSize;
+  }
+  else
+  {
+    visibleSize = outSize;
+    hiddenSize = inSize;
+  }
+  // Weight shape = k * d * n
+  weight = arma::cube(weights.memptr(), poolSize,
+      visibleSize, hiddenSize, false,
+      false);
+  // slabBias shape = k * n ==> diagMat(slabBias.col(i)) = k * k
+  slabBias = arma::mat(weights.memptr() + weight.n_elem,
+      poolSize, hiddenSize, false, false);
+
+  // lambdaBias shape = d * 1 ==> diagMat(lambdaBias.col(0)) = d * d
+  lambdaBias = arma::mat(weights.memptr() + weight.n_elem + slabBias.n_elem,
+      visibleSize, 1, false, false);
+
+  // spikeBias shape = 1 * N
+  spikeBias = arma::mat(weights.memptr() + weight.n_elem + slabBias.n_elem +
+      lambdaBias.n_elem, 1, hiddenSize, false, false);
+}
+
+template<typename InputDataType, typename OutputDataType>
+void SpikeSlabLayer<InputDataType, OutputDataType>::Sample(InputDataType&&
+    input, OutputDataType&& output)
+{
+  if (!typeVisible)
+  {
+    if (state.is_empty())
+      state = arma::zeros(visibleSize, 1);
+    Psgivenvh(std::move(state), std::move(input), std::move(mean),
+        std::move(output));
+    Pvgivensh(std::move(output), std::move(input), std::move(mean),
+        std::move(output));
+  }
+  else
+  {
+    for (size_t i = 0; i < hiddenSize; i++)
+    {
+      if (output.is_empty())
+        output = arma::zeros(hiddenSize, 1);
+      output(i) = LogisticFunction::Fn(arma::as_scalar(input.t() *
+          weight.slice(i).t() * arma::diagmat(slabBias.col(i)).i() *
+          weight.slice(i) * input + spikeBias.col(i)));
+      output(i) = math::RandBernoulli(output(i));
+    }
+  }
+}
+template<typename InputDataType, typename OutputDataType>
+void SpikeSlabLayer<InputDataType, OutputDataType>::Pvgivensh(arma::mat&& slab,
+    arma::mat&& hidden, arma::mat&& outMean, arma::mat&& output)
+{
+  variance = arma::diagmat(lambdaBias.col(0)).i();
+  mean = arma::zeros(variance.n_rows, 1);
+  for (size_t i = 0; i < weight.n_slices; i++)
+  {
+    mean += weight.slice(i).t() * slab.col(i) * arma::as_scalar(hidden(i));
+  }
+  mean = variance * mean;
+  GaussianDistribution dist(mean, variance);
```

Comment:
I suggest to replace that by `math::RandNormal()` since `GaussianDistribution` is too heavy.